### ISSUE TITLE
2.0.2 causing Out of Memory errors

### ISSUE THREAD

**Issue by timoschwarzer** [verbb/formie#993]
### Describe the bug

I just updated to Formie 2.0.2 and now all requests (page & admin) are returning an HTTP 500 after a few seconds. Looking at the logs, all requests fail due to being out of memory.

Downgrading to 2.0.1 fixes this problem.

### Steps to reproduce

1. Use Formie 2.0.2
2. ???

I'm unable to reproduce this in my local dev setup (which doesn't have a lot of content), but I can consistently reproduce it on our production server. I'm still investigating what exactly causes this and will update this issue accordingly.

### Form settings

N/A

### Craft CMS version

4.1.4.1

### Plugin version

2.0.2

### Multi-site?

Yes

### Additional context

```
[18-Jul-2022 13:45:46 Europe/Berlin] PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 528384 bytes) in /app/vendor/yiisoft/yii2/base/Component.php on line 209
[18-Jul-2022 13:45:46 Europe/Berlin] PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 163840 bytes) in /app/vendor/voku/portable-ascii/src/voku/helper/ASCII.php on line 475
[18-Jul-2022 13:45:46 Europe/Berlin] PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 20480 bytes) in /app/vendor/yiisoft/yii2/base/Component.php on line 209
[18-Jul-2022 13:45:46 Europe/Berlin] PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 20480 bytes) in /app/vendor/yiisoft/yii2/db/Command.php on line 1176
[18-Jul-2022 13:45:49 Europe/Berlin] PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 20480 bytes) in /app/vendor/yiisoft/yii2/base/Component.php on line 209
[18-Jul-2022 13:45:49 Europe/Berlin] PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 40960 bytes) in /app/vendor/twig/twig/src/Extension/CoreExtension.php on line 1544
[18-Jul-2022 13:45:50 Europe/Berlin] PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 528384 bytes) in /app/vendor/yiisoft/yii2/base/Component.php on line 209
[18-Jul-2022 13:45:50 Europe/Berlin] PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 20480 bytes) in /app/vendor/yiisoft/yii2/db/Command.php on line 1176
```

**Comment by engram-design** [comment]
Are you able to look at your logs in `/storage/logs` to see what might be the issue? It might help to archive all current logs (add them to a folder) then try and recreate the issue. If it's a 500 error, it should be logging the issue somewhere.

**Comment by timoschwarzer** [comment]
@engram-design The logs I posted are from `phperrors.log`. Since OOMs are fatal errors, it does not generate a stack trace.

**Comment by engram-design** [comment]
Hmm, I would've though even the `web.log` would give us a clue. I'll investigate what might be going on, but like you, nothing on development.

Are your PHP versions different between local and production?

**Comment by timoschwarzer** [comment]
> Are your PHP versions different between local and production?

No, both use this as a base image:

```dockerfile
FROM php:8.1-fpm-buster

RUN set -x && \
    apt-get -y update && \
    apt-get install -y --no-install-recommends wget gnupg2 software-properties-common && \
    echo "deb http://apt.postgresql.org/pub/repos/apt $(lsb_release -cs)-pgdg main" > /etc/apt/sources.list.d/pgdg.list && \
    wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add - && \
    apt-get -y update && \
    apt-get install -y --no-install-recommends \
      wait-for-it \
      libfreetype6-dev libjpeg62-turbo-dev \
      libmcrypt-dev libpng-dev libbz2-dev \
      libssl-dev autoconf \
      libmagickwand-dev \
      ca-certificates curl g++ libicu-dev \
      libzip-dev libpq-dev libonig-dev sudo \
      nginx supervisor postgresql-client-12 \
      poppler-utils && \
    \
    wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add - && \
    echo "deb http://apt.postgresql.org/pub/repos/apt/ buster-pgdg main" | tee  /etc/apt/sources.list.d/pgdg.list && \
    apt-get purge -y wget && \
    \
    docker-php-ext-install \
      ftp gettext mbstring opcache \
      shmop sockets sysvmsg sysvsem sysvshm \
      zip iconv pdo_pgsql intl bcmath && \
    \
    pecl install redis imagick && \
    \
    docker-php-ext-enable redis imagick intl bcmath && \
    \
    echo "upload_max_filesize = 1024M" > /usr/local/etc/php/php.ini && \
    echo "post_max_size = 1024M" >> /usr/local/etc/php/php.ini && \
    echo "max_execution_time = 300" >> /usr/local/etc/php/php.ini && \
    echo "memory_limit = 256M" >> /usr/local/etc/php/php.ini

COPY ./nginx.conf /etc/nginx/nginx.conf
```

**Comment by engram-design** [comment]
Thanks, I can't see any difference with PHP 8.1 on my end. It does seem strange there's an issue, given that 2.0.2 was [mostly](https://github.com/verbb/formie/compare/2.0.1...2.0.2) additions, new events and superficial fixes.

Will keep investigating.

**Comment by timoschwarzer** [comment]
As much as I want to help investigating, I can only do so in the night when the production site is not visited a lot. But I'll keep an eye on it.

**Comment by superflausch** [comment]
I'm having the same issue here after installing the latest update on my local dev env. Disabling the plugin resolves the issue.

This happens with PHP 8.0 as well as 8.1 for me. Getting the same error in the log. I also noticed a lot of warnings in the web.log, but not sure if those are related:

`2022-07-19 09:50:04 [web.WARNING] [craft\elements\db\ElementQuery::prepare] Element query executed before Craft is fully initialized. {"memory":71849376}`

This is also a multi-site setup. I have a similar site running on the same machine without a multi-site setup running fine with the latest version of the plugin. Could the issue be related to multi-site setups?

**Comment by engram-design** [comment]
I don't suppose you'd be able to send through your database @superflausch ? Understand if you're unable to, but I've also made sure to test on a multi-site. Sounds like we're hitting an infinite loop somewhere, although stepping through the changes, I'm not quite sure how.

**Comment by superflausch** [comment]
Let me clean it up and I think I can send it to you. I will send it through your support form. I hope that's fine.

**Comment by engram-design** [comment]
Thanks for that @superflausch looks like that was a specific set of circumstances! This had to do with an Agree field used in the form, which contained a Link mark, which _also_ contained a reference to another element like an entry. This was causing an infinite loop of trying to resolve the entry, while also processing the Agree field's settings, and around and around.

Fixed in [2.0.3](https://github.com/verbb/formie/releases/tag/2.0.3)

**Comment by timoschwarzer** [comment]
Wow, good find! Thank you for fixing it! :smile:
